### <font color=green>By:</font><font color=#296EAA> Mostafa Alaa Abd-Elaal Farghly</font>
### <font color=green>Project Repository:</font> https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning
-------------------------------------

# Results of AlexNet, Resnet50, and Resnet101 CNN in Detecting Covid-19 Using CT and X-ray Images

The Experiments were conducted using four public databases:

<table>
  <tr>
      <th><center>Number</center></th>
    <th><center>Dataset's link</center></th>
      <th><center>Modality</center></th>
      <th><center>Data Size (Number of Images)</center></th>
  </tr>
    
  <tr>
      <td><center>1</center></td>
    <td><center>https://github.com/UCSD-AI4H/COVID-CT</center></td>
    <td><center>CT</center></td>
    <td><center>746</center></td>
  </tr>
  
  <tr>
      <td><center>2</center></td>
    <td><center>https://github.com/RunwenHu/COVID-19</center></td>
    <td><center>CT</center></td>
    <td><center>802</center></td>
  </tr>
  
  <tr>
      <td><center>3</center></td>
    <td><center>https://www.kaggle.com/plameneduardo/sarscov2-ctscan-dataset</center></td>
    <td><center>CT</center></td>
    <td><center>2481</center></td>
  </tr>
  
  <tr>
      <td><center>4</center></td>
    <td><center>https://www.dropbox.com/s/7rjw6oet4za01op/CovidDataset-20200427T133042Z-001.zip?dl=0</center></td>
    <td><center>X-ray</center></td>
    <td><center>284</center></td>
  </tr>
  
</table>

Some or all of the datasets were separately used to train, validate, and test each of the three models (AlexNet, Resnet50, and Resnet101). Moreover, for the models trained by any of the three CT images datasets, the other two whole datasets were used to test the model.


## Models and Results
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;In this section, the standard model's results against the different datasets will be shown along with the results of the same model after customization if any were made.

### 1. AlexNet
In this model we used "Adam Optimizer", and used each of the four datasets separately to train and test the model.
##### Model Training Code: https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning/blob/main/AlexNet/AlexNet.ipynb

![image](AlexNet.png "AlexNet consists of 5 Convolutional Layers and 3 Fully Connected Layers")

#### 1.1. The Standard AlexNet Model With Dataset 1

![image](./AlexNet/AlexNet_ds_1.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.7224</li>
  <li>Validation Accuracy = 0.4746</li>
  <li>Testing Accuracy = 0.6010</li>
</ul>

#### 1.2. The Standard AlexNet Model With Dataset 2

![image](./AlexNet/AlexNet_ds_2.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.6931</li>
  <li>Validation Accuracy = 0.5051</li>
  <li>Testing Accuracy = 0.5055</li>
</ul>

#### 1.3. The Standard AlexNet Model With Dataset 3

![image](./AlexNet/AlexNet_ds_3.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.5104</li>
  <li>Validation Accuracy = 0.4821</li>
  <li>Testing Accuracy = 0.5040</li>
</ul>

#### 1.4. The Standard AlexNet Model With Dataset 4
With this dataset, data augmentation was used.

![image](./AlexNet/AlexNet_ds_4.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.5000</li>
  <li>Validation Accuracy = 0.5000</li>
</ul>

### observation:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AlexNet's model with "Adam" optimizer has showed quite poor results with all the datasets.

### 2. AlexNet
In this model we used "SGD" optimizer (lr=0.0001, momentum=0.9), and used each of the four datasets separately to train and test the model.
##### Model Training Code: https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning/blob/main/AlexNet/AlexNet-SGD.ipynb

#### 2.1. The Standard AlexNet Model With Dataset 1

![image](./AlexNet/AlexNet_SGD_ds_1.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 1.0000</li>
  <li>Validation Accuracy = 0.7034</li>
  <li>Testing Accuracy = 0.6601</li>
</ul>

#### 2.2. The Standard AlexNet Model With Dataset 2

![image](./AlexNet/AlexNet_SGD_ds_2.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 1.0000</li>
  <li>Validation Accuracy = 0.8788</li>
  <li>Testing Accuracy = 0.8791</li>
</ul>

#### 2.3. The Standard AlexNet Model With Dataset 3

![image](./AlexNet/AlexNet_SGD_ds_3.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 1.0000</li>
  <li>Validation Accuracy = 0.9305</li>
  <li>Testing Accuracy = 0.8387</li>
</ul>

#### 2.4. The Standard AlexNet Model With Dataset 4
With this dataset, data augmentation was used.

![image](./AlexNet/AlexNet_SGD_ds_4.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.9375</li>
  <li>Validation Accuracy = 0.9500</li>
</ul>

### observation:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"SGD" optimizer (lr=0.0001, momentum=0.9) has improved AlexNet's performance, but no satisfactory results yet except for dataset 4 (X-ray).

### 3. Customized AlexNet
In this model we used "SGD" optimizer (lr=0.0001, momentum=0.9), and used each of the four datasets separately to train and test the model.
###### Model Training Code: https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning/blob/main/AlexNet/AlexNet-Customized.ipynb

### Customization
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Here, a normalization layer was added at the input of the model, nine Batch Normalization layers were added to each layer, and three Dropout layers of rate 0.4 were added to the three fully connected layers.

#### 3.1. The Customized AlexNet Model With Dataset 1

![image](./AlexNet/AlexNet_cust_ds_1.png)

After 40 epochs:
<ul>
  <li>Training Accuracy = 0.9576</li>
  <li>Validation Accuracy = 0.7203</li>
  <li>Testing Accuracy = 0.6601</li>
</ul>

#### 3.2. The Customized AlexNet Model With Dataset 2

![image](./AlexNet/AlexNet_cust_ds_2.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.9886</li>
  <li>Validation Accuracy = 0.8990</li>
  <li>Testing Accuracy = 0.9001</li>
</ul>

#### 3.3. The Customized AlexNet Model With Dataset 3

![image](./AlexNet/AlexNet_cust_ds_3.png)

After 40 epochs:
<ul>
  <li>Training Accuracy = 0.9966</li>
  <li>Validation Accuracy = 0.9507</li>
  <li>Testing Accuracy = 0.9395</li>
</ul>

#### 3.4. The Customized AlexNet Model With Dataset 4
With this dataset, data augmentation was used.

![image](./AlexNet/AlexNet_cust_ds_4.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.9777</li>
  <li>Validation Accuracy = 0.5000</li>
</ul>

### observation:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This customization showed better results for datasets 2 and 3, while showed worst results for dataset 4 (X-ray). Nevertheless, the results for dataset 1 stayed nearly the same.

### Conculosions on AlexNet and on datasets
<ul>
    <li>Dataset 1 is not good enough for AlexNet's model.</li>
    <li>The Customized model demonestrated the best results.</li>
    <li>"SGD" optimizer works beter than "Adam" optimmizer for all the datasets</li>
    <li>A low learning is needed for the "SGD" optimizer (a learning rate of 0.0001 was used)</li>
</ul>

--------------------------------------------------------------------------------------------

#### Please note that from now on dataset 4 will be Ignored as it is verey small (only 284 images), and it has showed relatively satisfactory results with the AlexNet model with "SGD" optimizer.

### 4. ResNet 50
In this model we used "SGD" optimizer (lr=0.0001, momentum=0.9), and used the three CT datasets separately to train and test the model.
###### Model Training Code: https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning/blob/main/ResNet_50/ResNet%2050.ipynb

#### 4.1. The ResNet 50 Model With Dataset 1

![image](./ResNet_50/Resnet_50_ds_1.png)

After 80 epochs:
<ul>
  <li>Training Accuracy = 0.9576</li>
  <li>Validation Accuracy = 0.7373</li>
  <li>Testing Accuracy = 0.7635</li>
</ul>

#### 4.2. The ResNet 50 Model With Dataset 2

![image](./ResNet_50/Resnet_50_ds_2.png)

After 80 epochs:
<ul>
  <li>Training Accuracy = 0.9690</li>
  <li>Validation Accuracy = 0.8788</li>
  <li>Testing Accuracy = 0.9011</li>
</ul>

#### 4.3. The ResNet 50 Model With Dataset 3

![image](./ResNet_50/Resnet_50_ds_3.png)

After 80 epochs:
<ul>
  <li>Training Accuracy = 0.9659</li>
  <li>Validation Accuracy = 0.9260</li>
  <li>Testing Accuracy = 0.9395</li>
</ul>

### observation:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ResNet 50 improved the results on dataset 1, while showed nearly similar results to the customized AlexNet model.

------------------------

### 5. ResNet 101
In this model we used "SGD" optimizer (lr=0.0001, momentum=0.9), and used the three CT datasets separately to train and test the model.
###### Model Training Code: https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning/blob/main/ResNet_101/ResNet%20101.ipynb

#### 5.1. The ResNet 101 Model With Dataset 1

![image](./ResNet_101/Resnet_101_ds_1.png)

After 80 epochs:
<ul>
  <li>Training Accuracy = 0.9647</li>
  <li>Validation Accuracy = 0.7119</li>
  <li>Testing Accuracy = 0.7685</li>
</ul>

#### 5.2. The ResNet 101 Model With Dataset 2

![image](./ResNet_101/Resnet_101_ds_2.png)

After 80 epochs:
<ul>
  <li>Training Accuracy = 0.9755</li>
  <li>Validation Accuracy = 0.8788</li>
  <li>Testing Accuracy = 0.9121</li>
</ul>

#### 5.3. The ResNet 101 Model With Dataset 3

![image](./ResNet_101/Resnet_101_ds_3.png)

After 80 epochs:
<ul>
  <li>Training Accuracy = 0.9692</li>
  <li>Validation Accuracy = 0.9148</li>
  <li>Testing Accuracy = 0.9194</li>
</ul>

### observation:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ResNet 101 showed nearly similar results to ResNet 50.

## ResNet 101 Testing
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Now since ResNet 101 and ResNet 50 demonestrated the best results, we will take ResNet model through further testing with the different CT datasets.

First we will load the datasets

In [1]:
import tensorflow as tf

# Import Dataset 1
ds_1_CT = tf.keras.preprocessing.image_dataset_from_directory(r'./dataset_1_CT',
                                                              image_size=(227, 227),
                                                              batch_size=32,
                                                              label_mode='categorical')

# Import Dataset 2
ds_2_CT = tf.keras.preprocessing.image_dataset_from_directory(r'./dataset_2_CT',
                                                              image_size=(227, 227),
                                                              batch_size=32,
                                                              label_mode='categorical')

# Import Dataset 3
ds_3_CT = tf.keras.preprocessing.image_dataset_from_directory(r'./dataset_3_CT',
                                                              image_size=(227, 227),
                                                              batch_size=32,
                                                              label_mode='categorical')

Found 746 files belonging to 2 classes.
Found 802 files belonging to 2 classes.
Found 2481 files belonging to 2 classes.


<font color=green>Then we load the models of ResNet trained by the three CT datasets and test them.</font>

#### 1. Evaluate ResNet 101 trained by dataset 1

In [2]:
# Load ResNet model trained by dataset 1
Resnet_101_ds1 = tf.keras.models.load_model('../Resnet_101/Resnet_101_dataset_1')

In [3]:
# Evaluate against dataset 1
loss, acc = Resnet_101_ds1.evaluate(ds_1_CT, verbose=2)
print('ResNet 101 of dataset 1 against dataset 1, accuracy: {:5.2f}%'.format(100 * acc))

24/24 - 8s - loss: 0.3277 - accuracy: 0.8727
ResNet 101 of dataset 1 against dataset 1, accuracy: 87.27%


In [4]:
# Evaluate against dataset 2
loss, acc = Resnet_101_ds1.evaluate(ds_2_CT, verbose=2)
print('ResNet 101 of dataset 1 against dataset 2, accuracy: {:5.2f}%'.format(100 * acc))

26/26 - 9s - loss: 0.3135 - accuracy: 0.8716
ResNet 101 of dataset 1 against dataset 2, accuracy: 87.16%


In [5]:
# Evaluate against dataset 3
loss, acc = Resnet_101_ds1.evaluate(ds_3_CT, verbose=2)
print('ResNet 101 of dataset 1 against dataset 3, accuracy: {:5.2f}%'.format(100 * acc))

78/78 - 27s - loss: 0.8314 - accuracy: 0.5187
ResNet 101 of dataset 1 against dataset 3, accuracy: 51.87%


#### 2. Evaluate ResNet 101 trained by dataset 2

In [6]:
# Load ResNet model trained by dataset 2
Resnet_101_ds2 = tf.keras.models.load_model('../Resnet_101/Resnet_101_dataset_2')

In [7]:
# Evaluate against dataset 1
loss, acc = Resnet_101_ds2.evaluate(ds_1_CT, verbose=2)
print('ResNet 101 of dataset 2 against dataset 1, accuracy: {:5.2f}%'.format(100 * acc))

24/24 - 8s - loss: 0.2915 - accuracy: 0.8954
ResNet 101 of dataset 2 against dataset 1, accuracy: 89.54%


In [8]:
# Evaluate against dataset 2
loss, acc = Resnet_101_ds2.evaluate(ds_2_CT, verbose=2)
print('ResNet 101 of dataset 2 against dataset 2, accuracy: {:5.2f}%'.format(100 * acc))

26/26 - 9s - loss: 0.1482 - accuracy: 0.9564
ResNet 101 of dataset 2 against dataset 2, accuracy: 95.64%


In [9]:
# Evaluate against dataset 3
loss, acc = Resnet_101_ds2.evaluate(ds_3_CT, verbose=2)
print('ResNet 101 of dataset 2 against dataset 3, accuracy: {:5.2f}%'.format(100 * acc))

78/78 - 27s - loss: 0.9762 - accuracy: 0.4901
ResNet 101 of dataset 2 against dataset 3, accuracy: 49.01%


#### 3. Evaluate ResNet 101 trained by dataset 3

In [10]:
# Load ResNet model trained by dataset 3
Resnet_101_ds3 = tf.keras.models.load_model('../Resnet_101/Resnet_101_dataset_3')

In [11]:
# Evaluate against dataset 1
loss, acc = Resnet_101_ds3.evaluate(ds_1_CT, verbose=2)
print('ResNet 101 of dataset 3 against dataset 1, accuracy: {:5.2f}%'.format(100 * acc))

24/24 - 8s - loss: 1.5445 - accuracy: 0.5845
ResNet 101 of dataset 3 against dataset 1, accuracy: 58.45%


In [12]:
# Evaluate against dataset 2
loss, acc = Resnet_101_ds3.evaluate(ds_2_CT, verbose=2)
print('ResNet 101 of dataset 3 against dataset 2, accuracy: {:5.2f}%'.format(100 * acc))

26/26 - 9s - loss: 1.3208 - accuracy: 0.6559
ResNet 101 of dataset 3 against dataset 2, accuracy: 65.59%


In [13]:
# Evaluate against dataset 3
loss, acc = Resnet_101_ds3.evaluate(ds_3_CT, verbose=2)
print('ResNet 101 of dataset 3 against dataset 3, accuracy: {:5.2f}%'.format(100 * acc))

78/78 - 27s - loss: 0.1485 - accuracy: 0.9532
ResNet 101 of dataset 3 against dataset 3, accuracy: 95.32%


## Customized ALexNet Testing
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Now since AlexNet demonestrated approximately equal results to the results of ResNet 101 and Resnet 50, when using datasets 2 and 3, we will have further testing on those models with the different CT datasets.

<font color=green>Then we load the models of Customized AlexNet trained by the CT datasets 1 and 2 and test them.</font>

#### 1. Evaluate Customized AlexNet trained by dataset 2

In [14]:
# Load Customized AlexNet model trained by dataset 2
Cust_AlexNet_ds2 = tf.keras.models.load_model('../AlexNet/AlexNet_cust_dataset_2')

In [15]:
# Evaluate against dataset 1
loss, acc = Cust_AlexNet_ds2.evaluate(ds_1_CT, verbose=2)
print('Customized AlexNet of dataset 2 against dataset 1, accuracy: {:5.2f}%'.format(100 * acc))

24/24 - 2s - loss: 0.2495 - accuracy: 0.9276
Customized AlexNet of dataset 2 against dataset 1, accuracy: 92.76%


In [16]:
# Evaluate against dataset 2
loss, acc = Cust_AlexNet_ds2.evaluate(ds_2_CT, verbose=2)
print('Customized AlexNet of dataset 2 against dataset 2, accuracy: {:5.2f}%'.format(100 * acc))

26/26 - 3s - loss: 0.1663 - accuracy: 0.9763
Customized AlexNet of dataset 2 against dataset 2, accuracy: 97.63%


In [17]:
# Evaluate against dataset 3
loss, acc = Cust_AlexNet_ds2.evaluate(ds_3_CT, verbose=2)
print('Customized AlexNet of dataset 2 against dataset 3, accuracy: {:5.2f}%'.format(100 * acc))

78/78 - 8s - loss: 0.9369 - accuracy: 0.4458
Customized AlexNet of dataset 2 against dataset 3, accuracy: 44.58%


#### 2. Evaluate Customized AlexNet trained by dataset 3

In [18]:
# Load Customized AlexNet model trained by dataset 3
Cust_AlexNet_ds3 = tf.keras.models.load_model('../AlexNet/AlexNet_cust_dataset_3')

In [19]:
# Evaluate against dataset 1
loss, acc = Cust_AlexNet_ds3.evaluate(ds_1_CT, verbose=2)
print('Customized AlexNet of dataset 3 against dataset 1, accuracy: {:5.2f}%'.format(100 * acc))

24/24 - 2s - loss: 0.9968 - accuracy: 0.5563
Customized AlexNet of dataset 3 against dataset 1, accuracy: 55.63%


In [20]:
# Evaluate against dataset 2
loss, acc = Cust_AlexNet_ds3.evaluate(ds_2_CT, verbose=2)
print('Customized AlexNet of dataset 3 against dataset 2, accuracy: {:5.2f}%'.format(100 * acc))

26/26 - 3s - loss: 0.8767 - accuracy: 0.6185
Customized AlexNet of dataset 3 against dataset 2, accuracy: 61.85%


In [21]:
# Evaluate against dataset 3
loss, acc = Cust_AlexNet_ds3.evaluate(ds_3_CT, verbose=2)
print('Customized AlexNet of dataset 3 against dataset 3, accuracy: {:5.2f}%'.format(100 * acc))

78/78 - 8s - loss: 0.1178 - accuracy: 0.9851
Customized AlexNet of dataset 3 against dataset 3, accuracy: 98.51%


## Observations
<ul>
    <li>Testing those two models of Customized AlexNet trained by datasets 1 and 2, demonestrated beter results than the ResNet 101 model trained by the same datasets.</li>
    <li>The conflict between the datasets 1 and 2 and dataset 3 is still the same by this model.</li>
</ul>

## Conclusion
<ul>
    <li>Customized AlexNet Trained by either dataset 2 or 3, gives the best results among (Standard AlexNet, ResNet50 and ResNet 101)
<ul>

------------------------------------------

# Summary

<ul>
    <li>Dataset 4 (X-ray) is quite small and not good enough for training.</li>
    <li>AlexNet works beter with "SGD" optimizer rather than "Adam" optimizer.</li>
    <li> Customizing AlexNet by normalizing the input, adding batch normalization, and dropout layer helped in improving the results</li>
    <li>Even though the Customized AlexNet is much simpler and has much more less learning time than both the ResNet 50 and the ResNet 101, it still demonestrated beter results for detecting Covid-19.</li>
    <li>A conflict appears between datasets 1 and 2 and dataset 3. Nevertheless, it is quite plausible to mix all the data and observe the results when training Customized AlexNet and ResNet 101.</li>
<ul>
    

### Accuracies of Customized AlexNet vs ResNet 101 (When Trained By Dataset 2)

<table style="width:100%">    
    <tr>
        <th></th>
        <th><center>Customized AlexNet</center></th>
        <th><center>ResNet 101</center></th>
    </tr>
    <tr>
        <th><center>Validation Accuracy</center></th>
        <td><center>98.86%</center></td>
        <td><center>97.55%</center></td>
    </tr>
    <tr>
        <th><center>Test Accuracy (tested by 10% of dataset 2)</center></th>
        <td><center>90.01%</center></td>
        <td><center>91.21%</center></td>
    </tr>
    <tr>
        <th><center>Test by dataset 1 (the whole dataset)</center></th>
        <td><center>92.76%</center></td>
        <td><center>89.54%</center></td>
    </tr>
    <tr>
        <th><center>Test by dataset 2 (the whole dataset)</center></th>
        <td><center>97.63%</center></td>
        <td><center>95.64%</center></td>
    </tr>
    <tr>
        <th><center>Test by dataset 3 (the whole dataset)</center></th>
        <td><center>44.58%</center></td>
        <td><center>49.01%</center></td>
    </tr>
</table>

### Accuracies of Customized AlexNet vs ResNet 101 (When Trained By Dataset 3)

<table style="width:100%">    
    <tr>
        <th></th>
        <th><center>Customized AlexNet</center></th>
        <th><center>ResNet 101</center></th>
    </tr>
    <tr>
        <th><center>Validation Accuracy</center></th>
        <td><center>95.07%</center></td>
        <td><center>96.92%</center></td>
    </tr>
    <tr>
        <th><center>Test Accuracy (tested by 10% of dataset 3)</center></th>
        <td><center>93.95%</center></td>
        <td><center>91.94%</center></td>
    </tr>
    <tr>
        <th><center>Test by dataset 1 (the whole dataset)</center></th>
        <td><center>55.63%</center></td>
        <td><center>58.45%</center></td>
    </tr>
    <tr>
        <th><center>Test by dataset 2 (the whole dataset)</center></th>
        <td><center>61.85%</center></td>
        <td><center>65.59%</center></td>
    </tr>
    <tr>
        <th><center>Test by dataset 3 (the whole dataset)</center></th>
        <td><center>98.51%</center></td>
        <td><center>95.32%</center></td>
    </tr>
</table>

---------------------------------------
# <font color=red>Update</font>

## ResNet 101 and Customized AlexNet Results with mixed datasets
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Now we have mixed the all three CT datasets (1, 2, and 3), and trained our best two models with them.

### 1. Customized AlexNet
In this model we used "SGD" optimizer (lr=0.0001, momentum=0.9), and the three CT datasets mixed together to train and test the model (90% training and validation, 10% Testing).
###### Model Training Code: https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning/blob/main/AlexNet/AlexNet-CT-mix.ipynb

![image](./AlexNet/AlexNet_CT_mix.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.9864</li>
  <li>Validation Accuracy = 0.9380</li>
  <li>Testing Accuracy = 0.8967</li>
</ul>

### 1. ResNet 101
In this model we used "SGD" optimizer (lr=0.0001, momentum=0.9), and the three CT datasets mixed together to train and test the model (90% training and validation, 10% Testing).
###### Model Training Code: https://github.com/Mostafa-Farghly/Covid-19-Detection-With-Deep-Learning/blob/main/ResNet_101/ResNet-101-CT-mix.ipynb

![image](./ResNet_101/Resnet_101_ct_mix.png)

After 50 epochs:
<ul>
  <li>Training Accuracy = 0.9347</li>
  <li>Validation Accuracy = 0.8597</li>
  <li>Testing Accuracy = 0.8487</li>
</ul>

### Opservation:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Agian, the customized AlexNet model outperforms the ResNet 101 model.